lxd-agent/network.go

```diff
@@ -69,3 +81,157 @@ func serverTLSConfig() (*tls.Config, error) {
+		nicData[conf.MACAddress] = struct {
+			Name string
+			MTU  uint32
+		}{
+			Name: newInterfaceName,
+			MTU:  conf.MTU,
+		}
+	}
+
+	configureNIC := func(currentInterfaceName, currentMACAddress, currentMTU string) error {
+		reverter := revert.New()
+		defer reverter.Fail()
+
+		nic, ok := nicData[currentMACAddress]
+		if !ok {
+			return nil
+		}
+
+		_, err = shared.RunCommand("ip", "link", "set", "down", currentInterfaceName)
+		if err != nil {
+			return err
+		}
+		reverter.Add(func() {
+			shared.RunCommand("ip", "link", "set", "up", currentInterfaceName)
+		})
+
+		_, err = shared.RunCommand("ip", "link", "set", currentInterfaceName, "name", nic.Name)
+		if err != nil {
+			return err
+		}
+		reverter.Add(func() {
+			_, err = shared.RunCommand("ip", "link", "set", nic.Name, "name", currentInterfaceName)
+		})
+
+		if nic.MTU != 0 {
+			_, err = shared.RunCommand("ip", "link", "set", "dev", nic.Name, "mtu", strconv.Itoa(int(nic.MTU)))
+			if err != nil {
+				return err
+			}
+			reverter.Add(func() {
+				_, err = shared.RunCommand("ip", "link", "set", "dev", nic.Name, "mtu", currentMTU)
+			})
+		}
+
+		_, err = shared.RunCommand("ip", "link", "set", "up", nic.Name)
+		if err != nil {
+			return err
+		}
+
+		reverter.Success()
+		return nil
+	}
+
+	netDir := "/sys/class/net"
+	files, err := ioutil.ReadDir(netDir)
+	if err != nil {
+		return
+	}
+
+	for _, f := range files {
+		addrBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "address"))
+		if err != nil {
+			logger.Error("Unable to read MAC address", log15.Ctx{
+				"error": err,
+			})
+			continue
+		}
+		macAddress := string(bytes.TrimRight(addrBytes, "\n"))
+
+		// Parse MAC to ensure it is of the same form as written by the OS (avoiding casing errors).
+		hw, err := net.ParseMAC(macAddress)
+		if err != nil {
+			logger.Error("Unable to parse MAC address", log15.Ctx{
+				"error": err,
+			})
+			continue
+		}
+
+		mtuBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "mtu"))
+		if err != nil {
+			logger.Error("Unable to read MTU", log15.Ctx{
+				"error": err,
+			})
+			continue
+		}
+		mtuStr := string(bytes.TrimRight(mtuBytes, "\n"))
+
+		err = configureNIC(f.Name(), hw.String(), mtuStr)
+		if err != nil {
+			logger.Error("Unable to reconfigure network interface", log15.Ctx{
+				"error": err,
+			})
+		}
+	}
+
+	netplanBytes, err := ioutil.ReadFile("/etc/netplan/10-lxc.yaml")
+	if err == nil {
+		newNetplanBytes := bytes.Replace(netplanBytes, []byte("enp5s0"), []byte("eth0"), 1)
+		err := ioutil.WriteFile("/etc/netplan/10-lxc.yaml", newNetplanBytes, 0644)
+		logger.Error("New netplan config could not be applied", log15.Ctx{
+			"error": err,
+		})
+
+		_, err = shared.RunCommand("netplan", "apply")
+		logger.Error("New netplan config could not be applied", log15.Ctx{
+			"error": err,
+		})
+	} else if !os.IsNotExist(err) {
+		logger.Error("New netplan config could not be applied", log15.Ctx{
+			"error": err,
+		})
+	}
```

Comment:
@tomponline renaming `enp5s0` to `eth0` and calling `netplan apply` allows for this to work with default configurations. This won't work for the issue author however. We would need to apply some sensible default netplan config for each nic device.

Reply:
@stgraber was it your intention that the templates should be updated to support this feature?

Reply:
@markylaing https://github.com/lxc/lxd/issues/7233#issuecomment-616909079 suggests the template changes are out of scope for this PR (at least for now):

> though we wouldn't want it done by default as we need to ship default network configs in the images and so can't have varying interface names in there (just like we assume eth0 for containers).

Reply:
I'm still a little confused by this. I'm suggesting that the `lxd-agent` changes the netplan config inside the running instance only if `agent.rename_interfaces=true` (and only if the netplan file exists). 

So it wouldn't be changing the file in the image template, it would just be automating some network configuration, similar to how we are calling `ip link set ...`.

Some configuration would need to be added for each netplan `ethernet` as in the original issue description for this to work out-of-the-box (though I am not sure what that should be yet).

Also, I guess we should either edit the netplan, or use `ip link set ..`, but not both, depending on whether the netplan file exists or not.

Reply:
> I'm still a little confused by this. I'm suggesting that the `lxd-agent` changes the netplan config inside the running instance only if `agent.rename_interfaces=true` (and only if the netplan file exists)

I don't think we would want the lxd-agent modifying system config, especially as netplan is specific to Ubuntu, and we'd then also have to support all the various other types of network configuration systems (networkd, NetworkManager, `/etc/network/interfaces` etc).

Perhaps @stgraber has a view on this?

Reply:
As a more general example, today if you add an additional NIC to an instance, e.g. `eth1` for containers, then it won't be auto configured in netplan/systemd-networkd etc inside the container for DHCP. We don't modify an existing container's configuration currently.

Reply:
The way we generally modify the contents of files inside an instance is via cloud-init config.

Reply:
I have had a conversation with @stgraber and we wont be modifying any configuration. This means that the `agent.rename_interfaces` flag will most likely break networking for the instance without further configuration, but this is ok because the majority of users with a use case for this will be applying their own network configuration anyway.

That said, a simple way to use this without breaking the networking is:
```
$ lxc init images:ubuntu/20.04 v1 --vm -c agent.rename_interfaces=true
$ lxc config device override v1 eth0 name=enp5s0
$ lxc start v1
```

Reply:
Sounds good!